File: backend/dotnet/src/Mateu.Core/RouteRegistry.cs
using Mateu.Dtos;
using YamlDotNet.Serialization;

namespace Mateu.Core;

/// <summary>
/// One entry of a mount's route registry: what a URL resolves to. (Mirrors Java's
/// <c>io.mateu.uidl.data.RouteEntry</c> and Python's <c>RouteEntry</c>.)
/// </summary>
/// <param name="Route">Path RELATIVE to the mount, with <c>:name</c> segments for path parameters.
/// "" is the mount's root view.</param>
/// <param name="Definition">The layout file. Null when the view model supplies its own tree.</param>
/// <param name="ViewModel">The server class. <b>Optional on purpose</b>: a statically deployed route
/// has no server behind it, so a definition with no view model is a valid, complete route.</param>
/// <param name="FixedParams">Pinned — NOT overridable by the request. Re-applied on the server
/// rather than trusted from the client, or "fixed" would be a suggestion and flipping one via the
/// query string would be a capability escalation.</param>
/// <param name="DefaultParams">Seeded — the request may override them.</param>
/// <param name="Parent">The ABSOLUTE route of the screen whose slot this route fills, or null for a
/// top-level route. A sub-route with a parent does not replace the page: the parent renders its
/// shell and this route's screen is nested into the parent's slot. Set when the authored
/// <c>children</c> tree is flattened. (Mirrors Java's RouteEntry.parent.)</param>
/// <param name="Children">Sub-routes nested under this one, authored RELATIVE to it. Each child
/// fills this screen's slot. This is the AUTHORING shape; the registry flattens it into absolute
/// entries carrying <c>Parent</c>, so a table entry read at runtime has an empty Children.</param>
/// <param name="State">Literal values that seed the route's component/route state on entry (the
/// componentState scope). They enter at the client-state/defaults precedence level, so a pinned
/// parameter still wins. (Mirrors Java's RouteEntry.state.)</param>
/// <param name="AppState">Literal values that seed the app state on entry (the app-scoped store
/// that [AppContext] also feeds, persisted across navigation). (Mirrors RouteEntry.appState.)</param>
/// <param name="Data">The route's component/route data, a reference to a named data source; there
/// is no literal data channel, data is always sourced. Resolved when the route loads.</param>
/// <param name="AppData">The route's app data, a reference to a named data source resolved once at
/// app scope (shared across routes).</param>
public sealed record RouteEntry(
    string Route,
    string? Definition,
    string? ViewModel,
    IReadOnlyDictionary<string, object?> FixedParams,
    IReadOnlyDictionary<string, object?> DefaultParams,
    string? Parent = null,
    IReadOnlyList<RouteEntry>? Children = null,
    IReadOnlyDictionary<string, object?>? State = null,
    IReadOnlyDictionary<string, object?>? AppState = null,
    RestDataSourceDto? Data = null,
    RestDataSourceDto? AppData = null)
{
    public static RouteEntry Of(string route, string? viewModel) =>
        new(route, null, viewModel, EmptyParams, EmptyParams);

    internal static readonly IReadOnlyDictionary<string, object?> EmptyParams =
        new Dictionary<string, object?>();

    /// <summary>Whether this route fills the slot of a parent screen rather than replacing the page.</summary>
    public bool HasParent() => !string.IsNullOrWhiteSpace(Parent);

    public IReadOnlyList<string> PathParams() =>
        Route.Split('/')
            .Where(s => s.StartsWith(':') && s.Length > 1)
            .Select(s => s[1..])
            .ToList();

    /// <summary>Defaults first (with <c>State</c> seeded at that same level), then whatever the
    /// request brought, then the fixed ones — which win over everything, which is the whole point of
    /// declaring them fixed.</summary>
    public Dictionary<string, object?> ResolveParams(IReadOnlyDictionary<string, object?>? fromRequest)
    {
        var resolved = new Dictionary<string, object?>(DefaultParams);
        // `State` seeds at the defaults level too (literal component state the route brings).
        if (State is not null)
            foreach (var kv in State) resolved.TryAdd(kv.Key, kv.Value);
        if (fromRequest is not null)
            foreach (var kv in fromRequest) resolved[kv.Key] = kv.Value;
        foreach (var kv in FixedParams) resolved[kv.Key] = kv.Value;
        return resolved;
    }
}

/// <summary>A matched route: the entry that answered, and the path parameters read off the URL.</summary>
public sealed record RouteMatch(RouteEntry Entry, IReadOnlyDictionary<string, object?> PathParams)
{
    public Dictionary<string, object?> Params(IReadOnlyDictionary<string, object?>? incoming)
    {
        var fromRequest = new Dictionary<string, object?>();
        if (incoming is not null)
            foreach (var kv in incoming) fromRequest[kv.Key] = kv.Value;
        foreach (var kv in PathParams) fromRequest[kv.Key] = kv.Value;
        return Entry.ResolveParams(fromRequest);
    }
}

/// <summary>A mount's route table.</summary>
public sealed record RouteTable(IReadOnlyList<RouteEntry> Routes)
{
    public static readonly RouteTable Empty = new(new List<RouteEntry>());

    /// <summary>Authored entries replace derived ones outright rather than being combined field by
    /// field: a half-overridden route would be far harder to reason about than a replaced one.</summary>
    public RouteTable MergedOver(RouteTable derived)
    {
        var byRoute = new Dictionary<string, RouteEntry>();
        foreach (var entry in derived.Routes) byRoute[entry.Route] = entry;
        foreach (var entry in Routes) byRoute[entry.Route] = entry;
        return new RouteTable(byRoute.Values.ToList());
    }

    /// <summary>The entry answering a concrete path. Static routes are tried before parameterised
    /// ones, so <c>orders/new</c> is never swallowed by <c>orders/:id</c>; among parameterised
    /// matches the most specific wins, so matching does not depend on declaration order.</summary>
    public RouteMatch? Match(string? path)
    {
        var target = RouteRegistry.Normalize(path);
        var targetSegments = target.Length == 0 ? Array.Empty<string>() : target.Split('/');
        RouteMatch? best = null;
        foreach (var entry in Routes)
        {
            var pattern = RouteRegistry.Normalize(entry.Route);
            var patternSegments = pattern.Length == 0 ? Array.Empty<string>() : pattern.Split('/');
            if (patternSegments.Length != targetSegments.Length) continue;
            var pathParams = new Dictionary<string, object?>();
            var matched = true;
            for (var i = 0; i < patternSegments.Length; i++)
            {
                var segment = patternSegments[i];
                if (segment.StartsWith(':') && segment.Length > 1)
                    pathParams[segment[1..]] = targetSegments[i];
                else if (segment != targetSegments[i]) { matched = false; break; }
            }
            if (!matched) continue;
            if (best is null || entry.PathParams().Count < best.Entry.PathParams().Count)
                best = new RouteMatch(entry, pathParams);
        }
        return best;
    }
}

/// <summary>
/// Reads a mount's <c>routes.yaml</c> from the specs directory, next to the definitions it routes to.
///
/// <para>Why a registry and not just attributes: an attribute says "this class lives at this path",
/// the one-to-one case. An entry binds a definition, a view model and parameters independently, so
/// the same screen can answer several routes with different parameters pinned, and a route can exist
/// with no server class behind it at all.</para>
/// </summary>
public sealed class RouteRegistry
{
    public const string FileName = "routes.yaml";

    private static readonly IDeserializer Yaml = new DeserializerBuilder().Build();
    private readonly string _dir;
    private RouteTable? _authored;

    public RouteRegistry(string? dir = null)
        => _dir = dir ?? Environment.GetEnvironmentVariable("MATEU_SPECS_DIR")
                      ?? Path.Combine("specs", "ui");

    public RouteTable Authored() => _authored ??= Load();

    public RouteMatch? Match(string? path) => Authored().Match(path);

    /// <summary>An app a deployment contributes: a mount root, with the class that backs it (null for
    /// a purely-DSL app) and the definition it renders (null for a class-based app). (Mirrors Java's
    /// RouteRegistry.AppRef and Python's AppRef.)</summary>
    public sealed record AppRef(string Route, string? ClassName, string? Definition)
    {
        public bool IsDsl() => string.IsNullOrWhiteSpace(ClassName);
    }

    /// <summary>
    /// Every app of the deployment, from TWO producers merged into ONE table — the same "authored
    /// wins" rule the routes and sources use:
    /// <list type="bullet">
    /// <item><b>derived</b>: the reflection-discovered <c>[Ui]</c> classes (each an app at its path),
    /// supplied by the caller (there is no compile-time index here as in Java);</item>
    /// <item><b>authored</b>: the <c>type: UI</c> mounts discovered in the specs directory — which
    /// need NO class, so a deployment can ship a purely-DSL app and announce it here like any
    /// other.</item>
    /// </list>
    /// A mount at the same base path as a class replaces it (authored wins). This is the enumeration a
    /// federated shell (or a bundle/tool) reads to announce apps uniformly, with or without a class
    /// behind them.
    /// </summary>
    public IReadOnlyList<AppRef> Apps(IEnumerable<AppRef>? derived = null)
    {
        var byRoute = new Dictionary<string, AppRef>();
        if (derived is not null)
            foreach (var d in derived)
                byRoute[Normalize(d.Route)] = d with { Route = Normalize(d.Route) };
        foreach (var mount in ScanMounts())
            byRoute[mount.BasePath] = new AppRef(mount.BasePath, null, DefinitionOf(mount));
        return byRoute.Values.ToList();
    }

    /// <summary>A DSL mount discovered by content: a base path and the route files that make it up.</summary>
    private sealed record MountDescriptor(string BasePath, IReadOnlyList<string> RouteFiles);

    /// <summary>Scans the specs directory for <c>type: UI</c> files (by content, not by filename), so
    /// several DSL apps can coexist. (Mirrors Java's MountRegistry.mounts.)</summary>
    private IReadOnlyList<MountDescriptor> ScanMounts()
    {
        var mounts = new List<MountDescriptor>();
        if (!Directory.Exists(_dir)) return mounts;
        foreach (var file in Directory.EnumerateFiles(_dir, "*.*", SearchOption.AllDirectories)
                     .Where(f => f.EndsWith(".yaml") || f.EndsWith(".yml")))
        {
            try
            {
                if (Yaml.Deserialize<object?>(File.ReadAllText(file)) is not IDictionary<object, object> root)
                    continue;
                if (Str(root, "type") != "UI") continue;
                var basePath = Normalize(Str(root, "basePath") ?? Str(root, "base_path"));
                var routeFiles = new List<string>();
                if (root.TryGetValue("routes", out var routes))
                {
                    if (routes is IEnumerable<object> list)
                        routeFiles.AddRange(list.Select(n => n?.ToString()).Where(s => s is not null)!);
                    else if (routes is string one)
                        routeFiles.Add(one);
                }
                mounts.Add(new MountDescriptor(basePath, routeFiles));
            }
            catch
            {
                // a broken descriptor must not take app enumeration down — skip it.
            }
        }
        return mounts;
    }

    /// <summary>The definition bound to a DSL mount's ROOT route: load its route files (relative
    /// entries prefixed with the mount base path), match the base path, read the definition. That
    /// root entry is where a DSL app's shell lives — a route with a definition and no class.</summary>
    private string? DefinitionOf(MountDescriptor mount)
    {
        foreach (var routeFile in mount.RouteFiles)
        {
            var path = Path.Combine(_dir, routeFile);
            if (!File.Exists(path)) continue;
            try
            {
                if (Yaml.Deserialize<object?>(File.ReadAllText(path)) is not { } root) continue;
                var nodes = root switch
                {
                    IDictionary<object, object> map when map.TryGetValue("routes", out var r) => r as IEnumerable<object>,
                    IEnumerable<object> list => list,
                    _ => null,
                };
                if (nodes is null) continue;
                var entries = new List<RouteEntry>();
                foreach (var node in nodes)
                    if (node is IDictionary<object, object> entry)
                        FlattenNode(entry, parentRoute: null, prefix: "", basePath: mount.BasePath, entries);
                var definition = new RouteTable(entries).Match(mount.BasePath)?.Entry.Definition;
                if (!string.IsNullOrWhiteSpace(definition)) return definition;
            }
            catch
            {
                // ignore a broken route file — try the next one.
            }
        }
        return null;
    }

    private RouteTable Load()
    {
        var path = Path.Combine(_dir, FileName);
        if (!File.Exists(path)) return RouteTable.Empty;
        try
        {
            var root = Yaml.Deserialize<object?>(File.ReadAllText(path));
            // Both shapes are accepted: a bare list of entries, or a `routes:` envelope.
            var nodes = root switch
            {
                IDictionary<object, object> map when map.TryGetValue("routes", out var r) => r as IEnumerable<object>,
                IEnumerable<object> list => list,
                _ => null,
            };
            if (nodes is null) return RouteTable.Empty;
            // A standalone `type: Routes` file may tag itself with a `basePath` header (group/prefix),
            // so a class-declared mount authors its inner routes with no separate descriptor. Routes
            // are still relative here; the base path is applied when flattening. (Mirrors Java's
            // MountRegistry.routeFileMounts / RouteRegistry basePath prefixing.)
            var basePath = root is IDictionary<object, object> rootMap
                ? Normalize(Str(rootMap, "basePath") ?? Str(rootMap, "base_path"))
                : "";
            var entries = new List<RouteEntry>();
            foreach (var node in nodes)
            {
                if (node is not IDictionary<object, object> entry) continue;
                // Flatten this node and its nested `children` into flat entries, composing each
                // child's route relative to its parent and prefixing the base path.
                FlattenNode(entry, parentRoute: null, prefix: "", basePath: basePath, entries);
            }
            return new RouteTable(entries);
        }
        catch
        {
            // A broken routes.yaml must not take the app down: the attribute-declared routes still
            // work. Losing every route because of a syntax error in an optional file would be worse
            // than the problem the file solves.
            return RouteTable.Empty;
        }
    }

    /// <summary>Flattens one authored node and its nested <c>children</c> into flat entries. A
    /// child's route is composed RELATIVE to its parent (<c>orders</c> under <c>use-cases/rra</c>
    /// becomes <c>use-cases/rra/orders</c>) and carries the parent's route as <c>Parent</c>, so the
    /// sub-route renders into the parent screen's slot instead of replacing the page. The base path
    /// is prefixed onto every absolute route. (Mirrors Java's RouteRegistry.flattenNode + prefix.)</summary>
    private static void FlattenNode(
        IDictionary<object, object> node,
        string? parentRoute,
        string prefix,
        string basePath,
        List<RouteEntry> entries)
    {
        var relative = Normalize(Str(node, "route"));
        var full = prefix.Length == 0 ? relative
            : relative.Length == 0 ? prefix
            : prefix + "/" + relative;
        entries.Add(new RouteEntry(
            Prefix(basePath, full),
            Str(node, "definition"),
            Str(node, "viewModel") ?? Str(node, "view_model"),
            Params(node, "fixedParams", "fixed_params"),
            Params(node, "defaultParams", "default_params"),
            parentRoute is null ? null : Prefix(basePath, parentRoute),
            null,
            Params(node, "state", "state"),
            Params(node, "appState", "app_state"),
            DataSourceOf(node, "data"),
            DataSourceOf(node, "appData", "app_data")));
        if (node.TryGetValue("children", out var raw) && raw is IEnumerable<object> children)
            foreach (var child in children)
                if (child is IDictionary<object, object> childNode)
                    FlattenNode(childNode, full, full, basePath, entries);
    }

    /// <summary>Prefix a relative route with the mount base path:
    /// <c>(back-office, orders) → back-office/orders</c>.</summary>
    private static string Prefix(string basePath, string route)
    {
        if (string.IsNullOrEmpty(basePath)) return route;
        return string.IsNullOrEmpty(route) ? basePath : basePath + "/" + route;
    }

    /// <summary>Parses a <c>data</c>/<c>appData</c> node into a RestDataSourceDto. A bare string is
    /// the <c>data: countries</c> shorthand (a reference by name); an object may also carry inline
    /// endpoint fields. (Mirrors Java's RouteRegistry.dataSourceOf.)</summary>
    private static RestDataSourceDto? DataSourceOf(IDictionary<object, object> node, string key, string? altKey = null)
    {
        if (!node.TryGetValue(key, out var raw) && altKey is not null) node.TryGetValue(altKey, out raw);
        switch (raw)
        {
            case null:
                return null;
            case string s:
                return string.IsNullOrWhiteSpace(s) ? null : RestDataSourceDto.FromRef(s.Trim());
            case IDictionary<object, object> map:
                var refName = Str(map, "ref");
                return new RestDataSourceDto(Str(map, "url") ?? "")
                {
                    Ref = string.IsNullOrWhiteSpace(refName) ? null : refName,
                    Method = Str(map, "method"),
                    Body = Str(map, "body"),
                    ItemsPath = Str(map, "itemsPath") ?? Str(map, "items_path"),
                    ValuePath = Str(map, "valuePath") ?? Str(map, "value_path"),
                    LabelPath = Str(map, "labelPath") ?? Str(map, "label_path"),
                    Proxy = map.TryGetValue("proxy", out var p) && p is not null
                            && bool.TryParse(p.ToString(), out var b) && b,
                };
            default:
                return null;
        }
    }

    private static string? Str(IDictionary<object, object> node, string key) =>
        node.TryGetValue(key, out var v) && v is not null ? v.ToString() : null;

    private static IReadOnlyDictionary<string, object?> Params(
        IDictionary<object, object> node, string key, string altKey)
    {
        if (!node.TryGetValue(key, out var raw)) node.TryGetValue(altKey, out raw);
        if (raw is not IDictionary<object, object> map) return RouteEntry.EmptyParams;
        var values = new Dictionary<string, object?>();
        foreach (var kv in map)
        {
            var name = kv.Key?.ToString();
            if (name is not null) values[name] = kv.Value;
        }
        return values;
    }

    /// <summary>Routes are relative to the mount, so a leading or trailing slash carries no meaning.</summary>
    internal static string Normalize(string? route)
    {
        var r = route ?? "";
        var q = r.IndexOf('?');
        if (q >= 0) r = r[..q];
        r = r.Trim('/');
        return r is "_empty" or "_no_route" or "_no_home_route" ? "" : r;
    }
}

File: backend/python/mateu_core/route_registry.py
"""A mount's route registry — the Python mirror of Java's RouteEntry/RouteTable/RouteRegistry.

A *mount* is a UI application served at a base path (what ``@ui`` declares, whose decorated class is
the mount's root view). Everything inside it can be resolved through this registry, declared in a
``routes.yaml`` sitting next to the definitions it routes to.

Why a registry and not just decorators: a decorator says "this class lives at this path", the
one-to-one case. An entry binds a *definition*, a *view model* and *parameters* independently, so the
same screen can answer several routes with different parameters pinned, and a route can exist with no
server class behind it at all — which is what a statically deployed screen is.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml


def _normalize(route: str | None) -> str:
    """Routes are relative to the mount, so a leading or trailing slash carries no meaning."""
    r = (route or "").strip("/")
    return "" if r in ("_empty", "_no_route", "_no_home_route") else r


def _prefix(base_path: str, route: str) -> str:
    """Prefix a relative route with the mount base path: ``(shop, orders) → shop/orders``."""
    if not base_path:
        return route
    return base_path if not route else base_path + "/" + route


def _params_of(node: dict, *keys: str) -> dict[str, Any]:
    """A literal params/state map under any of the accepted keys (camelCase or snake_case)."""
    for key in keys:
        value = node.get(key)
        if isinstance(value, dict):
            return dict(value)
    return {}


def _data_source_of(node: dict, *keys: str) -> "RestSourceRef | None":
    """Parse a ``data``/``appData`` node into a :class:`RestSourceRef`. A bare string is the
    ``data: countries`` shorthand (a reference by name); an object may carry inline endpoint
    fields — mirrors Java's RouteRegistry.dataSourceOf."""
    for key in keys:
        if key not in node or node.get(key) is None:
            continue
        n = node[key]
        if isinstance(n, str):
            return RestSourceRef(ref=n)
        if isinstance(n, dict):
            return RestSourceRef(
                ref=n.get("ref"),
                url=n.get("url"),
                method=n.get("method"),
                body=n.get("body"),
                items_path=n.get("itemsPath") or n.get("items_path"),
                value_path=n.get("valuePath") or n.get("value_path"),
                label_path=n.get("labelPath") or n.get("label_path"),
                proxy=bool(n.get("proxy", False)),
            )
    return None


def _flatten_node(
    node: dict, parent_route: str | None, prefix: str, out: list["RouteEntry"]
) -> None:
    """Flatten one authored node and its nested ``children`` into flat entries. A child's route is
    composed RELATIVE to its parent (``orders`` under ``use-cases/rra`` → ``use-cases/rra/orders``)
    and carries the parent's route as :attr:`RouteEntry.parent`, so the sub-route renders into the
    parent screen's slot instead of replacing the page. Routes are still relative to the mount here;
    the base path is applied afterwards (mirrors Java's RouteRegistry.flattenNode)."""
    relative = _normalize(node.get("route"))
    full = relative if not prefix else (prefix if not relative else prefix + "/" + relative)
    out.append(
        RouteEntry(
            route=full,
            definition=node.get("definition"),
            view_model=node.get("viewModel") or node.get("view_model"),
            fixed_params=_params_of(node, "fixedParams", "fixed_params"),
            default_params=_params_of(node, "defaultParams", "default_params"),
            parent=parent_route,
            state=_params_of(node, "state"),
            app_state=_params_of(node, "appState", "app_state"),
            data=_data_source_of(node, "data"),
            app_data=_data_source_of(node, "appData", "app_data"),
        )
    )
    children = node.get("children")
    if isinstance(children, list):
        for child in children:
            if isinstance(child, dict):
                _flatten_node(child, parent_route=full, prefix=full, out=out)


def _with_base_path(entry: "RouteEntry", base_path: str) -> "RouteEntry":
    """Return the entry with its (and its parent's) route prefixed by the file's base path."""
    if not base_path:
        return entry
    from dataclasses import replace

    return replace(
        entry,
        route=_prefix(base_path, entry.route),
        parent=_prefix(base_path, entry.parent) if entry.has_parent() else entry.parent,
    )


@dataclass(frozen=True)
class RestSourceRef:
    """A route's ``data``/``appData`` source — a reference to a named entry of ``sources.yaml`` (the
    ``data: countries`` shorthand), or an inline endpoint. The Python mirror of Java's
    ``RestDataSource`` in this file (kept tiny: the registry only needs to carry the ref through)."""

    ref: str | None = None
    url: str | None = None
    method: str | None = None
    body: str | None = None
    items_path: str | None = None
    value_path: str | None = None
    label_path: str | None = None
    proxy: bool = False


@dataclass(frozen=True)
class RouteEntry:
    """What a URL resolves to.

    :param route: path relative to the mount, with ``:name`` segments for path parameters. ``""`` is
        the mount's root view.
    :param definition: the layout file. ``None`` when the view model supplies its own tree.
    :param view_model: the server class. **Optional on purpose**: a statically deployed route has no
        server behind it, so a definition with no view model is a valid, complete route.
    :param fixed_params: pinned — **not overridable by the request**. Re-applied on the server rather
        than trusted from the client, or "fixed" would be a suggestion and flipping one via the query
        string would be a capability escalation.
    :param default_params: seeded — the request may override them.
    :param parent: the ABSOLUTE route of the screen whose slot this route fills, or ``None`` for a
        top-level route. Set when the authored ``children`` tree is flattened.
    :param state: literal values seeding the route's component/route state on entry (at the defaults
        precedence level, so a pinned parameter still wins).
    :param app_state: literal values seeding the app-scope state on entry (merged UNDER the client's).
    :param data: the route's component/route data, a reference to a named source in ``sources.yaml``
        (there is no literal data channel; data is always sourced). Fetched when the route loads.
    :param app_data: the route's app-scope data, a reference to a named source resolved once at app
        scope (shared across routes).
    """

    route: str = ""
    definition: str | None = None
    view_model: str | None = None
    fixed_params: dict[str, Any] = field(default_factory=dict)
    default_params: dict[str, Any] = field(default_factory=dict)
    parent: str | None = None
    state: dict[str, Any] = field(default_factory=dict)
    app_state: dict[str, Any] = field(default_factory=dict)
    data: RestSourceRef | None = None
    app_data: RestSourceRef | None = None

    def has_parent(self) -> bool:
        """Whether this route fills the slot of a parent screen rather than replacing the page."""
        return bool(self.parent and self.parent.strip())

    def path_params(self) -> list[str]:
        return [s[1:] for s in self.route.split("/") if s.startswith(":") and len(s) > 1]

    def resolve_params(self, from_request: dict[str, Any] | None) -> dict[str, Any]:
        """Defaults first, then ``state`` at the same defaults level, then whatever the request
        brought, then the fixed ones — which win over everything, which is the whole point of
        declaring them fixed."""
        resolved = dict(self.default_params)
        for k, v in self.state.items():
            resolved.setdefault(k, v)
        resolved.update(from_request or {})
        resolved.update(self.fixed_params)
        return resolved


@dataclass(frozen=True)
class Match:
    entry: RouteEntry
    path_params: dict[str, Any]

    def params(self, incoming: dict[str, Any] | None) -> dict[str, Any]:
        from_request = dict(incoming or {})
        from_request.update(self.path_params)
        return self.entry.resolve_params(from_request)


@dataclass(frozen=True)
class RouteTable:
    routes: tuple[RouteEntry, ...] = ()

    @staticmethod
    def of(entries) -> "RouteTable":
        return RouteTable(tuple(entries or ()))

    def merged_over(self, derived: "RouteTable") -> "RouteTable":
        """Authored entries replace derived ones outright rather than being combined field by field:
        a half-overridden route would be far harder to reason about than a replaced one."""
        by_route: dict[str, RouteEntry] = {e.route: e for e in derived.routes}
        for entry in self.routes:
            by_route[entry.route] = entry
        return RouteTable(tuple(by_route.values()))

    def match(self, path: str | None) -> Match | None:
        """Static routes are tried before parameterised ones, so ``orders/new`` is never swallowed by
        ``orders/:id``; among parameterised matches the most specific wins, so matching does not
        depend on declaration order."""
        target = _normalize(path)
        target_segments = target.split("/") if target else []
        best: Match | None = None
        for entry in self.routes:
            pattern = _normalize(entry.route)
            pattern_segments = pattern.split("/") if pattern else []
            if len(pattern_segments) != len(target_segments):
                continue
            path_params: dict[str, Any] = {}
            matched = True
            for pattern_segment, target_segment in zip(pattern_segments, target_segments):
                if pattern_segment.startswith(":") and len(pattern_segment) > 1:
                    path_params[pattern_segment[1:]] = target_segment
                elif pattern_segment != target_segment:
                    matched = False
                    break
            if not matched:
                continue
            if best is None or len(entry.path_params()) < len(best.entry.path_params()):
                best = Match(entry, path_params)
        return best


@dataclass(frozen=True)
class AppRef:
    """An app a deployment contributes: a mount root, with the class that backs it (``None`` for a
    purely-DSL app) and the definition it renders (``None`` for a class-based app). The Python mirror
    of Java's ``RouteRegistry.AppRef`` and .NET's ``RouteRegistry.AppRef``."""

    route: str
    class_name: str | None = None
    definition: str | None = None

    def is_dsl(self) -> bool:
        """No server class backs it — it is announced purely from its ``type: UI`` mount."""
        return not (self.class_name and self.class_name.strip())


class RouteRegistry:
    """Reads ``routes.yaml`` from the specs directory, next to the definitions it routes to."""

    FILE = "routes.yaml"

    def __init__(self, directory: str | None = None) -> None:
        self._dir = Path(directory or os.environ.get("MATEU_SPECS_DIR") or Path("specs") / "ui")
        self._authored: RouteTable | None = None

    def authored(self) -> RouteTable:
        if self._authored is None:
            self._authored = self._load()
        return self._authored

    def match(self, path: str | None) -> Match | None:
        return self.authored().match(path)

    def apps(self, derived: "list[AppRef] | None" = None) -> list[AppRef]:
        """Every app of the deployment, from TWO producers merged into ONE table — the same
        "authored wins" rule the routes and sources use:

        * **derived**: the reflection-discovered ``@ui`` classes (each an app at its path), supplied
          by the caller (there is no compile-time index here as in Java);
        * **authored**: the ``type: UI`` mounts discovered in the specs directory — which need NO
          class, so a deployment can ship a purely-DSL app and announce it here like any other.

        A mount at the same base path as a class replaces it (authored wins). This is the enumeration
        a federated shell (or a bundle/tool) reads to announce apps uniformly, with or without a
        class behind them.
        """
        by_route: dict[str, AppRef] = {}
        for ref in derived or []:
            route = _normalize(ref.route)
            by_route[route] = AppRef(route, ref.class_name, ref.definition)
        for base_path, route_files in self._scan_mounts():
            by_route[base_path] = AppRef(
                base_path, None, self._definition_of(base_path, route_files)
            )
        return list(by_route.values())

    def _scan_mounts(self) -> list[tuple[str, list[str]]]:
        """Scan the specs directory for ``type: UI`` files (by content, not by filename), so several
        DSL apps can coexist. Returns each mount's base path + its listed route files. (Mirrors
        Java's MountRegistry.mounts.)"""
        mounts: list[tuple[str, list[str]]] = []
        if not self._dir.is_dir():
            return mounts
        for file in sorted(self._dir.rglob("*")):
            if file.suffix not in (".yaml", ".yml"):
                continue
            try:
                root = yaml.safe_load(file.read_text())
            except Exception:
                continue  # a broken descriptor must not take app enumeration down
            if not isinstance(root, dict) or root.get("type") != "UI":
                continue
            base_path = _normalize(root.get("basePath") or root.get("base_path"))
            routes = root.get("routes")
            if isinstance(routes, str):
                route_files = [routes]
            elif isinstance(routes, list):
                route_files = [r for r in routes if isinstance(r, str)]
            else:
                route_files = []
            mounts.append((base_path, route_files))
        return mounts

    def _definition_of(self, base_path: str, route_files: list[str]) -> str | None:
        """The definition bound to a DSL mount's ROOT route: load its route files (relative entries
        prefixed with the base path), match the base path, read the definition. That root entry is
        where a DSL app's shell lives — a route with a definition and no class."""
        for route_file in route_files:
            path = self._dir / route_file
            if not path.is_file():
                continue
            try:
                root = yaml.safe_load(path.read_text())
            except Exception:
                continue
            nodes = root.get("routes") if isinstance(root, dict) else root
            if not isinstance(nodes, list):
                continue
            relative: list[RouteEntry] = []
            for node in nodes:
                if isinstance(node, dict):
                    _flatten_node(node, parent_route=None, prefix="", out=relative)
            entries = [_with_base_path(e, base_path) for e in relative]
            match = RouteTable(tuple(entries)).match(base_path)
            if match is not None and match.entry.definition:
                return match.entry.definition
        return None

    def _load(self) -> RouteTable:
        path = self._dir / self.FILE
        if not path.is_file():
            return RouteTable()
        try:
            root = yaml.safe_load(path.read_text())
        except Exception:
            # A broken routes.yaml must not take the app down: the decorator-declared routes still
            # work, and the failure is loud in the log rather than fatal at boot.
            return RouteTable()
        if root is None:
            return RouteTable()
        # Both shapes are accepted: a bare list of entries, or a `routes:` envelope. A standalone
        # `type: Routes` envelope may tag itself with a `basePath` so a class-declared @ui("/shop")
        # mount authors its inner routes relatively and they resolve absolutely (last-wins on a
        # route collision is not needed for a single file — one file is one bucket).
        base_path = ""
        nodes = root
        if isinstance(root, dict):
            base_path = _normalize(root.get("basePath") or root.get("base_path"))
            nodes = root.get("routes")
        if not isinstance(nodes, list):
            return RouteTable()
        # Flatten each authored node (composing nested `children` relative to their parent and
        # setting `parent`), then prefix everything with the file's base path into absolute routes.
        relative: list[RouteEntry] = []
        for node in nodes:
            if isinstance(node, dict):
                _flatten_node(node, parent_route=None, prefix="", out=relative)
        entries = [_with_base_path(e, base_path) for e in relative]
        return RouteTable(tuple(entries))
